src/php/Command/ReplCommand.php

```diff
@@ -21,102 +20,125 @@ final class ReplCommand
-                if ($this->io->isBracketedPasteSupported()) {
-                    $this->io->output(self::DISABLE_BRACKETED_PASTE);
-                }
-
-                try {
-                    $this->checkInputAndAnalyze($input);
-                } catch (ExitException $e) {
-                    break;
-                }
+                $this->addLineFromPromptToBuffer();
+                $this->checkExitInputBuffer();
+                $this->analyzeInputBuffer();
+            } catch (ExitException $e) {
+                break;
             } catch (Throwable $e) {
+                $this->inputBuffer = [];
+                $this->io->output($this->style->red($e->getMessage() . PHP_EOL));
                 $this->io->output($e->getTraceAsString() . PHP_EOL);
             }
         }
 
-        $this->io->output($this->style->yellow("Bye!\n"));
+        $this->io->output($this->style->yellow('Bye!' . PHP_EOL));
     }
 
-    /**
-     * @throws ExitException
-     */
-    private function checkInputAndAnalyze(?string $input): void
+    private function addLineFromPromptToBuffer(): void
     {
-        if (null === $input || 'exit' === $input) {
-            throw new ExitException();
+        if ($this->io->isBracketedPasteSupported()) {
+            $this->io->output(self::ENABLE_BRACKETED_PASTE);
         }
 
-        if ('' === $input) {
-            return;
-        }
+        $isInitialInput = empty($this->inputBuffer);
+        $prompt = $isInitialInput ? self::INITIAL_PROMPT : self::OPEN_PROMPT;
+        $input = $this->io->readline($prompt);
 
-        $this->io->addHistory($input);
+        if ($this->io->isBracketedPasteSupported()) {
+            $this->io->output(self::DISABLE_BRACKETED_PASTE);
+        }
 
-        try {
-            $this->analyzeInput($input);
-        } catch (ParserException|ReaderException $e) {
-            $this->io->output($this->exceptionPrinter->getExceptionString($e, $e->getCodeSnippet()));
-        } catch (Throwable $e) {
-            $this->io->output($this->exceptionPrinter->getStackTraceString($e));
+        if ($input === null && $isInitialInput) {
+            // Ctrl+D will exit the repl
+            $this->inputBuffer[] = self::EXIT_REPL;
+        } elseif ($input === null && !$isInitialInput) {
+            // Ctrl+D will restart the buffer
+            $this->inputBuffer = [];
+        } else {
+            $this->inputBuffer[] = $input;
         }
     }
 
     /**
-     * @throws ReaderException
+     * @throws ExitException
      */
-    private function analyzeInput(string $input): void
+    private function checkExitInputBuffer(): void
+    {
+        $firstInput = $this->inputBuffer[0] ?? '';
+
+        if (self::EXIT_REPL === $firstInput) {
+            throw new ExitException();
+        }
+    }
+
+    private function analyzeInputBuffer(): void
     {
+        if ('' === end($this->inputBuffer)) {
+            array_pop($this->inputBuffer);
```

Comment:
This is to remove the latest introduced input when it's an empty string. This means, to remove the empty lines that you write while being filling the "inputBuffer" thing

Reply:
Ok, I see. But it is not required because the parser can still parse it (just whitespace). Why would you remove those empty lines? Maybe the user just wants to have empty lines.

Reply:
Because I didn't want to have them, actually 😁
Try it locally, and see it yourself, that you don't really want those empty lines in history. Because for the parser doesn't matter. They affect the history, and how it's saved. 
I took that decision based on the experience. Let me know what do you think. I could change it... but I think this is a good idea that improves the usability of the repl. 
I suggest keep it, and if we face some problems or we change our mind, we simply modify it later :)

Reply:
Ok cool. Then I will merge it now.